### ISSUE TITLE
ServiceLocator.Current.GetService(Type) throws EntryPointNotFoundException

### ISSUE THREAD

**Issue by ericwj** [PrismLibrary/Prism#1039]
### Problem
ServiceLocator.Current.GetService(Type) throws EntryPointNotFoundException

### Code review
`ServiceLocator.Current` is never used. Not by Prism itself.

I believe the IServiceLocator service can be added with a one-liner by app developers. It doesn't need to be included by default. Plus, it's broken as it stands. All references to types defined in CommonServiceLocator are a) trivial or almost trivial b) can be relaced by a single, DI container independent implementation and one abstract or virtual method and hence c) can be added after the fact with very little effort, or by offering a separate package to add it back.

### Package info

- Platform: UWP 10.0.15063
- Prism.Autofac version: 6.3.0 or any other package that has a (transitive) dependency on CommonServiceLocator 1.3.0
- UniversalWindows (>= 5.2.3) tried with 5.2.3 and 5.3.3

### Repro steps
Using Visual Studio 2017.3
Add new project, Universal Windows, Blank App MinVer 10.0.15063 Ver 10.0.15063.
Add NuGet Reference Prism.Autofac.
Change App.xaml.cs
```CS
using System.Threading.Tasks;
using Microsoft.Practices.ServiceLocation;
using Windows.ApplicationModel.Activation;

namespace CommonServiceLocatorCrash {
	sealed partial class App : Prism.Autofac.Windows.PrismAutofacApplication
    {
		protected override Task OnLaunchApplicationAsync(LaunchActivatedEventArgs args) {
			ServiceLocator.Current.GetService(typeof(Prism.Windows.Navigation.INavigationService));
			return Task.CompletedTask;
		}
	}
}
```
Change App.xaml
```XAML
<paw:PrismAutofacApplication
    x:Class="CommonServiceLocatorCrash.App"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:paw="using:Prism.Autofac.Windows"
    RequestedTheme="Light">

</paw:PrismAutofacApplication>
```
Set as Startup Project, F5
```
System.EntryPointNotFoundException: 'Entry point was not found.'
   at System.IServiceProvider.GetService(Type serviceType)
   at CommonServiceLocatorCrash.App.OnLaunchApplicationAsync(LaunchActivatedEventArgs args)
   at Prism.Windows.PrismApplication.<OnLaunched>d__53.MoveNext()
```

**Comment by brianlagunas** [comment]
Thanks for submitting the issue.  ServiceLocator is an anti-pattern, but I understand that a lot of developer like to use it.  We will talk about this.

**Comment by ericwj** [comment]
`IServiceProvider` isn't. This issue however breaks it.

**Comment by ericwj** [comment]
Forget that last remark. `IServiceProvider` works fine as long as you don't ask for it from `ServiceLocator.Current`.

How is `IServiceLocator` not supported with XF? Is it missing from the System namespace? Or doesn't it work with `CommonServiceLocator` 1.3.0 just like on UWP?

**Comment by brianlagunas** [comment]
The package will not add to a XF project due to an incompatible target profile.

**Comment by ericwj** [comment]
Which package? `IServiceProvider` is defined by the framework. Literally System.Runtime or mscorlib. I'd be surprised to find it isn't in every single project you can create with any profile on any .NET. Unfortunately I opted out of 30GB XF, so I can't just File New Project XF and type `System.` to find out

**Comment by brianlagunas** [comment]
The Microsoft.Extensions.DependencyInjection package. is this not what you were referring to?

**Comment by ericwj** [comment]
No, not really. It does make standardizing stuff a lot easier, but it is *not* required to start doing DI properly. If you're serious about doing DI properly and you want to do it yesterday, it's possible to standardize on `IServiceProvider` with a little bit of infrastructure in the Prism core PCL - which you can then remove once `Microsoft.Extensions.DependencyInjection` is everywhere.

It is just messy to have RegisterTypeIfMissing and the like. I'm surprised the code is as is if it in any way coming from Microsoft patterns and practices. I'm sure it had the ServiceLocator anti pattern from the start. They must have been learning a great deal since... I feel sorry looking around Prism.Windows honestly.

Have a look here https://developer.xamarin.com/api/type/System.IServiceProvider/

**Comment by brianlagunas** [comment]
As I mentioned, Prism.Windows did not have DI in it at all from the start.  This is not important enough to add any type of infrastructure to Prism to support this.  When all platforms support the necessary API's this can be revisited.  The current approach is the exact same as which was used for Prism for WPF which was written originally by the P&P team (this has splashes of ServiceLocator in it, but does not expect or recommend the end-user use that pattern).

**Comment by brianlagunas** [comment]
There are also a number of DI container features that must be supported by IServcieProvider before it can be used.  It's not as simple as a blanket GetService call.  Looking at the interface, there is no way this could be used as a standard across all platforms.

> you're serious about doing DI properly 
I'm sorry, but this comment cracks me up

**Comment by brianlagunas** [comment]
FYI, Prism.Windows has not really been touched since the P&P team first wrote it. It's been relatively the same since P&P first released it for Win8 :)

**Comment by ericwj** [comment]
Bullocks, didn't mean to offend. I've looked at Prism several times in several years. The past few weeks only for the first time a bit deeper, enabled by the fact that that it is now open source. My impression is you have taken compatibility very seriously and that's great. I'm just guessing you'd like to bump a few versions and get Prism where you really want it to be. And that that would include making DI a first-class principle.

I'm thinking about ServiceLocator, maybe it was separate, but I've come accross it using enterprise library I'm sure. I've used it. Don't recall exactly how I even got the idea, but it must've been in the docs at the very least.

In terms of support needed for IServiceProvider. It is already there in `Autofac.Extensions.DependencyInjection` and it is imho reasonable to expect Unity and SimpleInjector to implement a similar drop in DI extension - or risk being dropped. There are issues I've seen with Unity certianly in terms of how alive the project is and whether it'll remain viable in the future. About SimpleInjector I don't know.

But you are right, it'll certainly be a temporary fix and maybe not all that easy to try and fix this before .NET Standard is available. And after that it'll be fairly straightforward building on the infrastructure already there.

**Comment by ericwj** [comment]
> There are also a number of DI container features that must be supported by IServiceProvider before it can be used. It's not as simple as a blanket GetService call.

You mean resolution with parameters and things like that? In the Win10 solution, Prism defines its own providers to resolve types from strings and viewmodels from view types. I think all that's needed is to keep it that way. There will be `ActivatorUtilities` in `Microsoft.Extensions.DependencyInjection` to replace `Activator.CreateInstance` and light up DI in those providers.

**Comment by brianlagunas** [comment]
Once example would be runtime ctor parameter overrides. There is more to Prism than just Win10.  Actually Win10 is last on my list of priorities right now.

**Comment by ericwj** [comment]
I guess I'd have to put 30GB on my disk and take a look...

But really, if say Autofac can do that, then you can ask Autofac to do it with only an `IServiceProvider` in hand - at worst only after importing a namespace. If Prism does it, register a service and declare an extension method to use it in Prism.

**Comment by brianlagunas** [comment]
I'm not sure it's worth it or not.  In the end, you can still use your container of choice.  This is a very low priority for me as the benefit to my users does not stand out at me.  It just seems like a refactoring at what's here.  Using the IServiceProvider doesn't impact the "sharability" of VM's or business logic at all.

**Comment by dansiegel** [comment]
@ericwj you may want to consider Unity isn't maintained. As for SimpleInjector, Prism.Windows is the only platform using SimpleInjector at the moment, and to be blunt it's a pain in the *** to make it work for Prism. In my experience if you want to use an anti-pattern or really any pattern that isn't something SimpleInjector is built for you can expect long explanations about why what you want to do is wrong from SI's author.

**Comment by ericwj** [comment]
I don't care about any of them, I expect you'd want to keep them for compatibility reasons. Unless they can't follow along migrating to .NET Standard. Once that's there I'm fine not even using Autofac.

**Comment by ericwj** [comment]
Sharing code between Windows and XF is only possible with Shared Projects I guess. That's probably OK today, but tomorrow it will become sort of a hack. So, new up DelegateCommand in a VM? No, give me a factory from DI... And the right place to register it is in Prism.Windows.PrismApplication.

**Comment by brianlagunas** [comment]
ummm.... no.  That's not true at all.  I am sharing code between WPF, UWP, and Xamarin now.  There is no way I would ever use DI to new up an ICommand.

**Comment by ericwj** [comment]
Well there is no shared abstraction. So you must be using shared projects. Source code compatibility. Well it works. Maybe ICommand is too simple an example.

**Comment by brianlagunas** [comment]
I am not using shared projects. I actually refuse to use them for my own reasons.  I think the issue here is that you are not familiar enough with Prism as it is applies across all platforms.

**Comment by brianlagunas** [comment]
Back to the original issue, ServcieLocator is not intended to be supported out of the box.  Prism does not recommend using it, and if needed a developer can add support for it.  Also, it should be noted that not all containers use CommonServcieLocator, which would make doing this inconsistent with other containers.

**Comment by ericwj** [comment]
I in fact advocated for it to be removed in the original issue.

**Comment by brianlagunas** [comment]
Perfect! We are on the same page then :)  CommonServiceLocator is only added because they are dependencies of the container in question.  Prism does not utilize that reference.  Prism for UWP needs a complete review as it hasn't been touched since P&P wrote it.  When I review what's there, then the DI will be reevaluated.
